pimd/pim6_mroute.c

```diff
@@ -0,0 +1,1125 @@
+		}
+
+		pim_register_send((uint8_t *)buf + sizeof(struct ip6_hdr),
+				  ntohs(ip6_hdr->ip6_plen) - IPV6_HEADER_LENGTH,
+				  pim_ifp->primary_address, rpg, 0, up);
+	}
+	return 0;
+}
+
+static int pim_mroute_msg_wrongmif(int fd, struct interface *ifp,
+				   const struct mrt6msg *msg)
+{
+	struct pim_ifchannel *ch;
+	struct pim_interface *pim_ifp;
+	pim_sgaddr sg;
+
+	memset(&sg, 0, sizeof(pim_sgaddr));
+	IPV6_ADDR_COPY (&sg.src, &msg->im6_src);
+	IPV6_ADDR_COPY (&sg.grp, &msg->im6_dst);
+
+	/*
+	  Send Assert(S,G) on iif as response to WRONGMIF kernel upcall.
+
+	  RFC 4601 4.8.2.  PIM-SSM-Only Routers
+
+	  iif is the incoming interface of the packet.
+	  if (iif is in inherited_olist(S,G)) {
+	  send Assert(S,G) on iif
+	  }
+	*/
+
+	if (!ifp) {
+		if (PIM_DEBUG_MROUTE)
+			zlog_debug(
+				"%s: WRONGMIF (S,G)=%s could not find input interface for input_mif_index=%d",
+				__func__, pim_str_sg_dump(&sg), msg->im6_mif);
+		return -1;
+	}
+
+	pim_ifp = ifp->info;
+	if (!pim_ifp) {
+		if (PIM_DEBUG_MROUTE)
+			zlog_debug(
+				"%s: WRONGMIF (S,G)=%s multicast not enabled on interface %s",
+				__func__, pim_str_sg_dump(&sg), ifp->name);
+		return -2;
+	}
+
+	ch = pim_ifchannel_find(ifp, &sg);
+	if (!ch) {
+		pim_sgaddr star_g = sg;
+		if (PIM_DEBUG_MROUTE)
+			zlog_debug(
+				"%s: WRONGMIF (S,G)=%s could not find channel on interface %s",
+				__func__, pim_str_sg_dump(&sg), ifp->name);
+
+		star_g.src = in6addr_any;
+		ch = pim_ifchannel_find(ifp, &star_g);
+		if (!ch) {
+			if (PIM_DEBUG_MROUTE)
+				zlog_debug(
+					"%s: WRONGMIF (*,G)=%s could not find channel on interface %s",
+					__func__, pim_str_sg_dump(&star_g),
+					ifp->name);
+			return -3;
+		}
+	}
+
+	/*
+	  RFC 4601: 4.6.1.  (S,G) Assert Message State Machine
+
+	  Transitions from NoInfo State
+
+	  An (S,G) data packet arrives on interface I, AND
+	  CouldAssert(S,G,I)==TRUE An (S,G) data packet arrived on an
+	  downstream interface that is in our (S,G) outgoing interface
+	  list.  We optimistically assume that we will be the assert
+	  winner for this (S,G), and so we transition to the "I am Assert
+	  Winner" state and perform Actions A1 (below), which will
+	  initiate the assert negotiation for (S,G).
+	*/
+
+	if (ch->ifassert_state != PIM_IFASSERT_NOINFO) {
+		if (PIM_DEBUG_MROUTE) {
+			zlog_debug(
+				"%s: WRONGMIF (S,G)=%s channel is not on Assert NoInfo state for interface %s",
+				__func__, ch->sg_str, ifp->name);
+		}
+		return -4;
+	}
+
+	if (!PIM_IF_FLAG_TEST_COULD_ASSERT(ch->flags)) {
+		if (PIM_DEBUG_MROUTE) {
+			zlog_debug(
+				"%s: WRONGMIF (S,G)=%s interface %s is not downstream for channel",
+				__func__, ch->sg_str, ifp->name);
+		}
+		return -5;
+	}
+
+	if (assert_action_a1(ch)) {
+		if (PIM_DEBUG_MROUTE) {
+			zlog_debug(
+				"%s: WRONGMIF (S,G)=%s assert_action_a1 failure on interface %s",
+				__func__, ch->sg_str, ifp->name);
+		}
+		return -6;
+	}
+
+	return 0;
+}
+
+static int pim_mroute_msg(struct pim_instance *pim, const char *buf,
```

Comment:
Don't keep the function names same if its not required. Since this is being used only in this file, rename to suit v6.